### PR TITLE
add parachain MultiLocation

### PR DESCRIPTION
**Pull request by lana-shanghai**

### CODE REVIEW

File: src/MultiLocationEncoder.sol [hunk 1]
```diff
@@ -1,3 +1,43 @@
...
+        XCMEnums.XcmV3Junction firstJunction = XCMEnums.XcmV3Junction.Parachain;
+        return abi.encodePacked(
+            ScaleCodec.encodeU8(uint8(parents)),
+            ScaleCodec.encodeU8(uint8(interiorJunctions)),
+            ScaleCodec.encodeU8(uint8(firstJunction)),
+            CompactTypes.encodeCompactUint(parachainId)
+        );
+    }
+
+    function encodeMultiLocationSameConsensusAccountId32(
```
Comment (bidzyyys) [inline src/MultiLocationEncoder.sol:25]: IMHO it should be `X1(AccountId32)`:
```
Example: /AccountID32(0x234234)
```

**Review by bidzyyys — CHANGES_REQUESTED** [review]
Add `X1(AccountId32)`

**Review by sbp-parity — APPROVED** [review]
LGTM 👍

**Review by bidzyyys — APPROVED** [review]
LGTM 👍

File: src/MultiLocationEncoder.sol [hunk 2]
```diff
@@ -1,3 +1,54 @@
...
+contract MultiLocationEncoder {
+
+    function encodeMultiLocationSameConsensusParachain(
+        uint8 parents,
+        uint256 parachainId
+    )
+    public pure returns (bytes memory) {
+        XCMEnums.XcmV3Junctions interiorJunctions = XCMEnums.XcmV3Junctions.X1;
+        XCMEnums.XcmV3Junction firstJunction = XCMEnums.XcmV3Junction.Parachain;
+        return abi.encodePacked(
```
Comment (vgeddes) [inline src/MultiLocationEncoder.sol:17]: Should rather use [byte.concat](https://docs.soliditylang.org/en/v0.8.17/types.html#the-functions-bytes-concat-and-string-concat), which was introduced for this purpose. The former `abi.encodePacked` is [discouraged](https://github.com/ethereum/solidity/issues/10903) for concatenating bytes.

File: src/XCMEnums.sol [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        Polkadot, 
+        Kusama, 
+        Westend, 
+        Rococo,
+        Wococo,
+        Ethereum,
+        BitcoinCore, 
+        BitcoinCash
+    }
+}
```
Comment (vgeddes) [inline src/XCMEnums.sol:43]: needs newline 😄

File: src/XCMEnums.sol [hunk 4]
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+library XCMEnums {
```
Comment (vgeddes) [inline src/XCMEnums.sol:4]: This library is a bit unnecessary. Can define the enums at the toplevel like so:

```solidity
// SPDX-License-Identifier: Apache-2.0
pragma solidity ^0.8.19;

enum XcmV3Junction {
    Parachain,
    AccountId32,
    AccountIndex64,
    AccountKey20,
    PalletInstance,
    GeneralIndex,
    GeneralKey,
    OnlyChild,
    Plurality,
    GlobalConsensus
}

```

File: src/utils/ScaleCodec.sol [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    function encodeU16(uint16 input) internal pure returns (bytes2) {
+        return bytes2(reverse16(input));
+    }
+
+    function encodeU8(uint8 input) internal pure returns (bytes1) {
+        return bytes1(input);
+    }
+}
```
Comment (vgeddes) [inline src/utils/ScaleCodec.sol:111]: missing newline

File: src/MultiLocationEncoder.sol [hunk 6]
```diff
@@ -1,3 +1,54 @@
...
 pragma solidity ^0.8.13;
 
-contract MultiLocationEncoder {}
+import "./XCMEnums.sol";
+import "./utils/ScaleCodec.sol";
+import "./utils/CompactTypes.sol";
+
+contract MultiLocationEncoder {
+
+    function encodeMultiLocationSameConsensusParachain(
```
Comment (vgeddes) [inline src/MultiLocationEncoder.sol:10]: Is a fairly verbose function name. Given that name of the contract suggests multilocation encoding, could condense the name to `encodeRelativeParachain`.

**Review by vgeddes — COMMENTED** [review]
Looks great! Added a few minor comments.

**Pull request merged**